hkmc2/shared/src/test/mlscript/logicsub/popl22.mls

```diff
@@ -0,0 +1,298 @@
+//│ ║  l.24: 	f as ((Int -> Int) & (Str -> Str), Int | Str) -> Int | Str
+//│ ║        	^
+//│ ╟── because: cannot constrain  ('x1, 'x2) ->{'eff} 'res  <:  ((Int -> Int) & (Str -> Str), Int | Str) -> (Int | Str)
+//│ ╟── because: cannot constrain  Int | Str  <:  'x2
+//│ ╟── because: cannot constrain  Str  <:  'x2
+//│ ╙── because: cannot constrain  Str  <:  ¬(¬{Int})
+//│ ╔══[ERROR] Type error in reference with expected type ((Int -> Int) & (Str -> Str), Int | Str) ->{⊥} Int | Str
+//│ ║  l.24: 	f as ((Int -> Int) & (Str -> Str), Int | Str) -> Int | Str
+//│ ║        	^
+//│ ╟── because: cannot constrain  ('x1, 'x2) ->{'eff} 'res  <:  ((Int -> Int) & (Str -> Str), Int | Str) -> (Int | Str)
+//│ ╟── because: cannot constrain  Int | Str  <:  'x2
+//│ ╟── because: cannot constrain  Str  <:  'x2
+//│ ╟── because: cannot constrain  Str  <:  'app
+//│ ╟── because: cannot constrain  Str  <:  'app
+//│ ╟── because: cannot constrain  'app1  <:  Str
+//│ ╟── because: cannot constrain  'app1  <:  ¬(¬{Str})
+//│ ╙── because: cannot constrain  Int  <:  ¬(¬{Str})
+//│ ╔══[ERROR] Type error in reference with expected type ((Int -> Int) & (Str -> Str), Int | Str) ->{⊥} Int | Str
+//│ ║  l.24: 	f as ((Int -> Int) & (Str -> Str), Int | Str) -> Int | Str
+//│ ║        	^
+//│ ╟── because: cannot constrain  ('x1, 'x2) ->{'eff} 'res  <:  ((Int -> Int) & (Str -> Str), Int | Str) -> (Int | Str)
+//│ ╟── because: cannot constrain  Int | Str  <:  'x2
+//│ ╟── because: cannot constrain  Str  <:  'x2
+//│ ╟── because: cannot constrain  Str  <:  'app
+//│ ╟── because: cannot constrain  Str  <:  'app
+//│ ╟── because: cannot constrain  'x2  <:  Str
+//│ ╟── because: cannot constrain  'x2  <:  ¬(¬{Str})
+//│ ╙── because: cannot constrain  Int  <:  ¬(¬{Str})
+//│ ╔══[ERROR] Type error in reference with expected type ((Int -> Int) & (Str -> Str), Int | Str) ->{⊥} Int | Str
+//│ ║  l.24: 	f as ((Int -> Int) & (Str -> Str), Int | Str) -> Int | Str
+//│ ║        	^
+//│ ╟── because: cannot constrain  ('x1, 'x2) ->{'eff} 'res  <:  ((Int -> Int) & (Str -> Str), Int | Str) -> (Int | Str)
+//│ ╟── because: cannot constrain  Int | Str  <:  'x2
+//│ ╟── because: cannot constrain  Str  <:  'x2
+//│ ╟── because: cannot constrain  {0: 'x2}  <:  {0: Str}
+//│ ╟── because: cannot constrain  'x2  <:  Str
+//│ ╟── because: cannot constrain  'x2  <:  ¬(¬{Str})
+//│ ╙── because: cannot constrain  Int  <:  ¬(¬{Str})
+//│ Type: ((Int -> Int) & (Str -> Str), Int | Str) ->{⊥} Int | Str
+
+
+data class Prod[A, B](val fst: A, val snd: B)
+//│ Type: ⊤
+
+class
+  Tru()
+  Fls()
+  Z()
+//│ Type: ⊤
+
+let tru = new Tru()
+let fls = new Fls()
+//│ Type: ⊤
+
+fun prod(p, q) = new Prod(p, q)
+//│ Type: ⊤
+
+fun y: Any -> Tru | Fls
+//│ Type: ⊤
+
+prod(y(x => x), y(x => x)) as Prod[Tru, Tru] | Prod[Fls, Fls]
+//│ Type: Prod[Tru, Tru] | Prod[Fls, Fls]
+
+:e
+{ fst: y(x => x), snd: y(x => x) } as [ fst: Tru, snd: Tru ] | [ fst: Fls, snd: Fls ]
```

Comment:
Any reason why you use a record on the LHS and tuples on the RHS? Why are the RHS' treated like records? They should be one-element arrays. Remember that `[ fst: Tru, snd: Tru ]` is syntax sugar for `[ (fst: Tru, snd: Tru) ]`.